### ISSUE TITLE
Cannot derivate abs

### ISSUE THREAD

**Issue by eliemichel** [symengine/symengine#1696]
The expected result of `diff(abs(x), x)` is `sign(x)`, but it is not resolved correctly:

```python
>>> from symengine import Symbol, diff
>>> x = Symbol('x')
>>> expr = diff(abs(x), x)
>>> expr.subs({x: 1})
Derivative(1, 1)
```

Values in this `Derivative(Real, Real)` state then lead to random segfaults on further operations.

**Comment by rikardn** [comment]
Sympy ends up with the same ```Derivative(1, 1)``` although it doesn't crash.

The derivative of ```abs(x)``` is not defined at ```x=0``` so there might currently be no good way of representing this result. The same goes for the sign function.

**Comment by eliemichel** [comment]
I agree that it is not well defined for `x = 0`, but Sympy handles it differently:

```python
>>> import sympy as sym
>>> x = sym.Symbol('x', real=True)
>>> sym.diff(abs(x), x)
sign(x)
>>> sym.diff(sym.sign(x), x)
2*DiracDelta(x)
```

Also, concretely I'd like to be sure that using `expr.subs(valuation)` with a valuation containing all the free variables of `expr` reduces to a float, how do I do so?

**Comment by rikardn** [comment]
I can comfirm the above with ```real=True```.

```python
import sympy
x = sympy.Symbol('x')
sympy.diff(abs(x), x)
```

gives 

```
(re(x)*Derivative(re(x), x) + im(x)*Derivative(im(x), x))*sign(x)/x
```

and

```python
sympy.diff(sympy.sign(x), x)
```

gives

```
Derivative(sign(x), x)
```

and, this looked weird at first:

```
import symengine
x = symengine.Symbol('x')
sympy.diff(abs(x), x)
```

gives

```
Derivative(abs(x), x)
```

(this was actually what I tested first). This must be the derivative calculated by symengine.


The valuation must be buggy. It seems as if the value simply gets substituted into the derivative even though it hasn't been calculated.

**Comment by asmeurer** [comment]
There is a lot of good mathematical discussion about this here https://github.com/sympy/sympy/issues/8502.
